datastreams/windower/tumbling.go

```diff
@@ -0,0 +1,104 @@
+package windower
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/elastiflow/pipelines/datastreams/internal/partition"
+	"github.com/elastiflow/pipelines/datastreams/internal/pipes"
+)
+
+type status struct {
+	started bool
+	mu      sync.Mutex
+}
+
+func newStatus() *status {
+	return &status{
+		started: false,
+	}
+}
+
+func (t *status) running() bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return t.started
+}
+
+func (t *status) set(state bool) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.started = state
+}
+
+// tumbling buffers items for windowDuration once the first item arrives.
+// After windowDuration, it publishes the batch and clears it, then waits
+// for the next item to start a new timer.
+type tumbling[T any] struct {
+	*partition.Base[T]
+	ctx            context.Context
+	status         *status
+	windowDuration time.Duration
+}
+
+// newTumbling constructs a tumbling window partition. The first Push
+// after a flush will start a new window timer.
+func newTumbling[T any](
+	ctx context.Context,
+	out pipes.Senders[[]T],
+	errs chan<- error,
+	windowDuration time.Duration,
+) partition.Partition[T] {
+	if windowDuration <= 0 {
+		panic("interval must be > 0")
+	}
+
+	t := &tumbling[T]{
+		Base:           partition.NewBase[T](out, errs),
+		windowDuration: windowDuration,
+		ctx:            ctx,
+		status:         newStatus(),
+	}
+	return t
+}
+
+// Push adds to the current batch; if no timer is running, start one.
+func (t *tumbling[T]) Push(item T) {
+	t.Base.Push(item)
+
+	if !t.status.running() {
+		t.status.set(true)
+		go t.waitAndFlush()
+	}
```

Comment:
I'd use `tryStart` to atomically check and set the running status.

```suggestion
	if t.status.tryStart() {
		go t.waitAndFlush()
	}
```